presto-main/src/main/java/com/facebook/presto/sql/planner/iterative/rule/AddNotNullFiltersToJoinNode.java

```diff
@@ -0,0 +1,254 @@
+    {
+        if (joinFilter instanceof SpecialFormExpression && ((SpecialFormExpression) joinFilter).getForm() == AND) {
+            //If the RowExpression is an AND clause
+            List<RowExpression> arguments = ((SpecialFormExpression) joinFilter).getArguments();
+            if (arguments.size() == 2) {
+                RowExpression possibleMarker = arguments.get(0);
+                //And the fist argument is our "marker", we have already applied this rule to this JoinNode
+                return NOT_NULL_MARKER_TRUE_EXPRESSION == possibleMarker;
+            }
+        }
+
+        return false;
+    }
+
+    private RowExpression buildUpdatedJoinFilter(RowExpression existingFilter, RowExpression notNullFilter)
+    {
+        //We build the updated JoinFilter with a custom "marker"
+        //We use this "marker" to determine if we have already applied this rule in #hasRuleAlreadyApplied
+        return and(NOT_NULL_MARKER_TRUE_EXPRESSION, and(existingFilter, notNullFilter));
+    }
+
+    private Collection<VariableReferenceExpression> extractNotNullVariables(List<JoinNode.EquiJoinClause> joinCriteria, RowExpression joinFilter,
+            List<VariableReferenceExpression> candidates, Session session)
+    {
+        RowExpression combinedFilter = joinFilter;
+        for (JoinNode.EquiJoinClause criteria : joinCriteria) {
+            combinedFilter = and(combinedFilter, criteria.getLeft());
+            combinedFilter = and(combinedFilter, criteria.getRight());
+        }
+
+        return intersection(ImmutableSet.copyOf(candidates), inferNotNullVariables(combinedFilter, session));
+    }
+
+    private ImmutableSet<VariableReferenceExpression> inferNotNullVariables(RowExpression expression, Session session)
+    {
+        ImmutableSet.Builder<VariableReferenceExpression> builder = ImmutableSet.builder();
+        expression.accept(new ExtractInferredNotNullVariablesVisitor(functionAndTypeManager, session), builder);
+        return builder.build();
+    }
+
+    @VisibleForTesting
+    public static class ExtractInferredNotNullVariablesVisitor
+            extends DefaultRowExpressionTraversalVisitor<ImmutableSet.Builder<VariableReferenceExpression>>
+    {
+        private final FunctionAndTypeManager functionAndTypeManager;
+        private final Session session;
+
+        public ExtractInferredNotNullVariablesVisitor(FunctionAndTypeManager functionAndTypeManager, Session session)
+        {
+            this.functionAndTypeManager = functionAndTypeManager;
+            this.session = session;
+        }
+
+        @Override
+        public Void visitCall(CallExpression call, ImmutableSet.Builder<VariableReferenceExpression> context)
+        {
+            final FunctionHandle functionHandle = call.getFunctionHandle();
+            final FunctionMetadata functionMetadata = functionAndTypeManager.getFunctionMetadata(functionHandle);
+
+            switch (getNotNullInferenceStrategy(session)) {
```

Comment:
My concern is not on how to check if function can be applied on NULL input, but this idea of extracting variables from these functions to get not null check. And the additional guard is to be applied to whether to extract these variables.

Reply:
Resolving since we were discussing this [on the other thread](https://github.com/prestodb/presto/pull/18938#discussion_r1132773007)